widgets/sparkline/sparkline.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sparkline is a widget that draws a graph showing a series of values as vertical bars.
+package sparkline
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"sync"
+
+	"github.com/mum4k/termdash/canvas"
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/draw"
+	"github.com/mum4k/termdash/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// SparkLine draws a graph showing a series of values as vertical bars.
+//
+// Bars can have sub-cell height. The graphs scale adjusts dynamically based on
+// the largest visible value.
+//
+// Implements widgetapi.Widget. This object is thread-safe.
+type SparkLine struct {
+	// data are the data points the SparkLine displays.
+	data []int
+
+	// mu protects the SparkLine.
+	mu sync.Mutex
+
+	// opts are the provided options.
+	opts *options
+}
+
+// New returns a new SparkLine.
+func New(opts ...Option) *SparkLine {
+	opt := newOptions()
+	for _, o := range opts {
+		o.set(opt)
+	}
+	return &SparkLine{
+		opts: opt,
+	}
+}
+
+// Draw draws the SparkLine widget onto the canvas.
+// Implements widgetapi.Widget.Draw.
+func (sl *SparkLine) Draw(cvs *canvas.Canvas) error {
+	sl.mu.Lock()
+	defer sl.mu.Unlock()
+
+	ar := sl.area(cvs)
+	visible, max := visibleMax(sl.data, ar.Dx())
+	var curX int
+	if len(visible) < ar.Dx() {
```

Comment:
I'm a bit troubled by this statement. What if the data points given are more sparse (not for every cell)? Will checking the length still make sense?
If the input only takes sequential data points, maybe it should be specified.